src/chain/rpc.rs

```diff
@@ -0,0 +1,724 @@
+const BLOCK_NONCE_ERROR: &str = "failed to fetch an account nonce by the scanner client";
+
+const CHARGE_ASSET_TX_PAYMENT: &str = "ChargeAssetTxPayment";
+
+// Pallets
+
+const SYSTEM: &str = "System";
+const BALANCES: &str = "Balances";
+const UTILITY: &str = "Utility";
+const ASSETS: &str = "Assets";
+const BABE: &str = "Babe";
+const TRANSFER: &str = "Transfer";
+
+// Runtime APIs
+
+const AURA: &str = "AuraApi";
+
+pub async fn subscribe_blocks(client: &WsClient) -> Result<Subscription<BlockHead>, ErrorChain> {
+    Ok(client
+        .subscribe(
+            "chain_subscribeFinalizedHeads",
+            rpc_params![],
+            "chain_unsubscribeFinalizedHeads",
+        )
+        .await?)
+}
+
+pub async fn get_value_from_storage(
+    client: &WsClient,
+    whole_key: &str,
+    block: &BlockHash,
+) -> Result<Value, ErrorChain> {
+    let value: Value = client
+        .request(
+            "state_getStorage",
+            rpc_params![whole_key, block.to_string()],
+        )
+        .await?;
+    Ok(value)
+}
+
+pub async fn get_keys_from_storage(
+    client: &WsClient,
+    prefix: &str,
+    storage_name: &str,
+    block: &BlockHash,
+) -> Result<Value, ErrorChain> {
+    let keys: Value = client
+        .request(
+            "state_getKeys",
+            rpc_params![
+                format!(
+                    "0x{}{}",
+                    hex::encode(twox_128(prefix.as_bytes())),
+                    hex::encode(twox_128(storage_name.as_bytes()))
+                ),
+                block.to_string()
+            ],
+        )
+        .await?;
+    Ok(keys)
+}
+
+/// fetch genesis hash, must be a hexadecimal string transformable into
+/// H256 format
+pub async fn genesis_hash(client: &WsClient) -> Result<BlockHash, ErrorChain> {
+    let genesis_hash_request: Value = client
+        .request(
+            "chain_getBlockHash",
+            rpc_params![Value::Number(Number::from(0u8))],
+        )
+        .await
+        .map_err(ErrorChain::Client)?;
+    match genesis_hash_request {
+        Value::String(x) => BlockHash::from_str(&x),
+        _ => return Err(ErrorChain::GenesisHashFormat),
+    }
+}
+
+/// fetch block hash, to request later the metadata and specs for
+/// the same block
+pub async fn block_hash(
+    client: &WsClient,
+    number: Option<String>,
+) -> Result<BlockHash, ErrorChain> {
+    let rpc_params = if let Some(a) = number {
+        rpc_params![a]
+    } else {
+        rpc_params![]
+    };
+    let block_hash_request: Value = client
+        .request("chain_getBlockHash", rpc_params)
+        .await
+        .map_err(ErrorChain::Client)?;
+    match block_hash_request {
+        Value::String(x) => BlockHash::from_str(&x),
+        _ => return Err(ErrorChain::BlockHashFormat),
+    }
+}
+
+/// fetch metadata at known block
+pub async fn metadata(
+    client: &WsClient,
+    block: &BlockHash,
+) -> Result<RuntimeMetadataV15, ErrorChain> {
+    let metadata_request: Value = client
```

Comment:
```suggestion
    let block_str = format!("0x{}", block.to_string());
    let metadata_request: Value = client
```

Reply:
no, I've made a special interface for it that does all dangerous conversions! If we want to always append `0x`, we should do it there. Is it a real recommendation now? Because upstream tends to break these things as we know.